graphql/client.go

```diff
@@ -0,0 +1,134 @@
+package graphql
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"math/rand"
+	"net/http"
+	purl "net/url"
+	"time"
+
+	pauth "github.com/pinpt/go-common/auth"
+	"github.com/pinpt/go-common/hash"
+	pjson "github.com/pinpt/go-common/json"
+)
+
+// Client interface to client object
+type Client interface {
+	// Mutate makes _mutate_ http request to the GraphQL server
+	Mutate(query string, variables Variables, out interface{}) error
+	// Mutate makes _query_ http request to the GraphQL server
+	Query(query string, variables Variables, out interface{}) error
+}
+
+// Variables a simple alias to map[string]interface{}
+type Variables map[string]interface{}
+
+// NewClient returns a new instance of Client
+func NewClient(customerID string, userID string, apikey string, url string) (Client, error) {
+	headers, err := defaultHeaders(customerID, userID, apikey)
+	if err != nil {
+		return nil, err
+	}
+	u, err := purl.Parse(url)
+	if err != nil {
+		return nil, err
+	}
+	u.Path = "/graphql"
+	return &client{
+		headers: headers,
+		url:     u.String(),
+	}, nil
+}
+
+func defaultHeaders(customerID string, userID string, apikey string) (map[string]string, error) {
+	headers := map[string]string{}
+	headers["x-api-key"] = apikey
+	if customerID != "" {
+		var header struct {
+			CustomerID string `json:"customer_id,omitempty"`
+			UserID     string `json:"user_id,omitempty"`
+		}
+		header.CustomerID = customerID
+		header.UserID = userID
+		auth, err := pauth.EncryptString(pjson.Stringify(header), apikey)
+		if err != nil {
+			return nil, err
+		}
+		headers["x-api-customer"] = auth
+	}
+	return headers, nil
+}
+
+type client struct {
+	headers map[string]string
+	url     string
+}
+
+// Mutate makes _mutate_ http request to the GraphQL server
+func (c *client) Mutate(query string, variables Variables, out interface{}) error {
+	return c.do(query, variables, out)
+}
+
+// Mutate makes _query_ http request to the GraphQL server
+func (c *client) Query(query string, variables Variables, out interface{}) error {
+	return c.do(query, variables, out)
+}
+
+func (c *client) do(query string, variables Variables, out interface{}) error {
+
+	payload := struct {
+		Variables Variables `json:"variables"`
+		Query     string    `json:"query"`
+	}{variables, query}
+
+	data, err := json.Marshal(payload)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, c.url, bytes.NewBuffer(data))
```

Comment:
need to use the api package yo construct so we can get benefit of tls cert checking